### PR TITLE
Implement NSInvocationOperation

### LINKED ISSUE #2454: finish NSInvocationOperation stubs

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
The tests included herein will fail until #2490 is fixed.

This pull request also includes a small fix for NSException's reference platform compliance.

Fixes #2454.

### CODE REVIEW

File: Frameworks/Foundation/NSException.mm [hunk 1]
```diff
@@ -195,6 +195,10 @@ - (void)dealloc {
     [super dealloc];
 }
 
+- (NSString*)description {
```
Comment (aballway) [inline Frameworks/Foundation/NSException.mm]: 👍

File: Frameworks/Foundation/NSInvocationOperation.mm [hunk 2]
```diff
@@ -14,30 +14,131 @@
...
 - (instancetype)initWithTarget:(id)target selector:(SEL)sel object:(id)arg {
-    UNIMPLEMENTED();
-    return StubReturn();
+    NSMethodSignature* methodSignature = [target methodSignatureForSelector:sel];
+    if (!methodSignature) {
+        [self release];
+        return nil;
+    }
+
+    if (methodSignature.numberOfArguments > 3) {
```
Comment (aballway) [inline Frameworks/Foundation/NSInvocationOperation.mm]: nit: add a comment about why 3.  Threw me through a loop

File: Frameworks/Foundation/NSInvocationOperation.mm [hunk 3]
```diff
@@ -14,30 +14,131 @@
...
+                     __PRETTY_FUNCTION__,
+                     sel_getName(sel),
+                     methodSignature.numberOfArguments - 2);
+        return nil;
+    }
+
+    NSInvocation* invocation = [NSInvocation invocationWithMethodSignature:methodSignature];
+    invocation.target = target;
+    invocation.selector = sel;
+    if (methodSignature.numberOfArguments > 2) {
```
Comment (aballway) [inline Frameworks/Foundation/NSInvocationOperation.mm]: nit: can just do == 3 here since > 3 has been removed as an option

File: tests/unittests/Foundation/NSOperationTests.mm [hunk 4]
```diff
@@ -1367,4 +1367,4 @@ - (void)setReady:(BOOL)ready {
                         },
                         2),
                     "Operation did not finish in time.");
-}
\ No newline at end of file
+}
```
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSOperationTests.mm]: will to revert

File: Frameworks/Foundation/NSInvocationOperation.mm [hunk 5]
```diff
@@ -14,30 +14,131 @@
...
+
+        if (_caughtException) {
+            [_caughtException raise];
+        }
+
+        if (_invocation.methodSignature.methodReturnLength == 0) {
+            [NSException raise:NSInvocationOperationVoidResultException format:@"%hs: %@ has a void result", __PRETTY_FUNCTION__, self];
+        }
+
+        return [[_result retain] autorelease];
```
Comment (aballway) [inline Frameworks/Foundation/NSInvocationOperation.mm]: nit: why is this necessary?

File: tests/unittests/Foundation/NSOperationTests.mm (hunk 4 above)
Comment (aballway) [inline tests/unittests/Foundation/NSOperationTests.mm]: pls to revert

File: Frameworks/Foundation/NSInvocationOperation.mm (hunk 5 above)
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSInvocationOperation.mm]: `_result` is a member. If we didn't retain+autorelease, this would be potentially fatal:

```objc
id r = [operation result];
[operation release]; // remove last reference
[r anything]; // boom: last owning reference was in operation
```

File: Frameworks/Foundation/NSInvocationOperation.mm (hunk 2 above)
Comment (ms-jihua) [inline Frameworks/Foundation/NSInvocationOperation.mm]: fwiw it's because target and selector are the first two arguments in a NSMethodSignature

File: Frameworks/Foundation/NSInvocationOperation.mm [hunk 6]
```diff
@@ -14,30 +14,131 @@
...
-    return StubReturn();
+- (instancetype)initWithInvocation:(NSInvocation*)invocation {
+    if (!invocation) {
+        [self release];
+        return nil;
+    }
+
+    if (self = [super init]) {
+        _invocation = [invocation retain];
+        [_invocation retainArguments];
```
Comment (ms-jihua) [inline Frameworks/Foundation/NSInvocationOperation.mm]: why do we call this in this path but not the other one?

File: tests/unittests/Foundation/NSInvocationOperationTests.mm [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+- (id)methodWithObjectReturn {
+    return @"Hello";
+}
+
+- (id)methodWithObjectReturnAndArgument:(id)argument {
+    return argument;
+}
+
+- (unsigned long long)methodWithUnsignedLongLongReturn {
+    return 0xDEADBEEFCAFECAFEULL;
```
Comment (ms-jihua) [inline tests/unittests/Foundation/NSInvocationOperationTests.mm]: ...

File: Frameworks/Foundation/NSInvocationOperation.mm (hunk 6 above)
Comment (DHowett-MSFT) [inline Frameworks/Foundation/NSInvocationOperation.mm]: The other path goes through this one!

File: tests/unittests/Foundation/NSInvocationOperationTests.mm (hunk 7 above)
Comment (DHowett-MSFT) [inline tests/unittests/Foundation/NSInvocationOperationTests.mm]: Sorry t.t

File: Frameworks/Foundation/NSInvocationOperation.mm (hunk 6 above)
Comment (ms-jihua) [inline Frameworks/Foundation/NSInvocationOperation.mm]: whoops

File: tests/unittests/Foundation/NSInvocationOperationTests.mm [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+//****************************************************************************** 
+
+#import <TestFramework.h>
+#import <Foundation/Foundation.h>
+
+static NSString* _NSInvocationOperationTargetException = @"_NSInvocationOperationTargetException";
+
+namespace {
+struct _LargeStruct {
+    unsigned long long a, b, c, d;
```
Comment (aballway) [inline tests/unittests/Foundation/NSInvocationOperationTests.mm]: nit: implicit int

File: tests/unittests/Foundation/NSInvocationOperationTests.mm [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+
+- (id)methodWithObjectReturn {
+    return @"Hello";
+}
+
+- (id)methodWithObjectReturnAndArgument:(id)argument {
+    return argument;
+}
+
+- (unsigned long long)methodWithUnsignedLongLongReturn {
```
Comment (aballway) [inline tests/unittests/Foundation/NSInvocationOperationTests.mm]: _ut supra_
